### PR TITLE
feat(builtin): allow bundling ESM output with the pkg_npm rule

### PR DESCRIPTION
**Pull request by SkaillZ**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #2647 

## What is the new behavior?

An attribute has been added that allows generating ESM files.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by SkaillZ** [comment]
@googlebot I signed it!

File: internal/pkg_npm/pkg_npm.bzl [hunk 1]
```diff
@@ -98,6 +98,10 @@ PKG_NPM_ATTRS = dict(NODE_CONTEXT_ATTRS, **{
         doc = """Other targets which produce files that should be included in the package, such as `rollup_bundle`""",
         allow_files = True,
     ),
+    "include_esm": attr.bool(
```
Comment (alexeagle) [inline internal/pkg_npm/pkg_npm.bzl:101]: this is great, do you know why we need this flag though? we don't have flags for the other JS flavors. Would something go wrong if we just made this always true for everyone in the next release?

**Review by alexeagle — COMMENTED** [review]
thanks!

File: internal/pkg_npm/pkg_npm.bzl (hunk 1 above)
Comment (SkaillZ) [reply internal/pkg_npm/pkg_npm.bzl]: Sorry for the late reply! I don't think it would break anything so I've removed the flag.

**Review by alexeagle — APPROVED** [review]
thanks!

File: internal/pkg_npm/pkg_npm.bzl [hunk 2]
```diff
@@ -295,11 +295,15 @@ def _pkg_npm(ctx):
         # Only collect DefaultInfo files (not transitive)
         deps_files_depsets.append(dep.files)
 
-        # All direct & transitive JavaScript-producing deps
-        if JSModuleInfo in dep:
-            deps_files_depsets.append(dep[JSModuleInfo].sources)
+        # All direct and transitive deps that produce CommonJS modules
+        if JSNamedModuleInfo in dep:
```
Comment (alexeagle) [inline internal/pkg_npm/pkg_npm.bzl:299]: I'm concerned about dropping `JsModuleInfo` - I'm pretty sure it's possible that we get a dep that has JS files only in that provider but not in either of these two you've added. I think we should have all three. Would be nice if there was a red test result on this PR making me more confident...

Comment (SkaillZ) [reply internal/pkg_npm/pkg_npm.bzl:303]: I've added the `JSModuleInfo` back, but I'm not sure where this provider is created. If you could give me an example of a rule where it's used, I could add a test case with it as well.

**Review by alexeagle — APPROVED** [review]
thanks!

**Pull request merged**